server/protocols/webadmin/webadmin-data-jmap/src/main/java/org/apache/james/webadmin/data/jmap/MessageFastViewProjectionCorrector.java

```diff
@@ -0,0 +1,142 @@
+                                       MessageFastViewPrecomputedProperties.Factory projectionItemFactory) {
+        this.usersRepository = usersRepository;
+        this.mailboxManager = mailboxManager;
+        this.messageFastViewProjection = messageFastViewProjection;
+        this.projectionItemFactory = projectionItemFactory;
+    }
+
+    Mono<Void> correctAllProjectionItems(Progress progress) {
+        try {
+            return Iterators.toFlux(usersRepository.list())
+                .concatMap(username -> correctUsersProjectionItems(progress, username))
+                .then();
+        } catch (UsersRepositoryException e) {
+            return Mono.error(e);
+        }
+    }
+
+    private Mono<Void> correctUsersProjectionItems(Progress progress, Username username) {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(username);
+            return Flux.fromIterable(mailboxManager.search(MailboxQuery.privateMailboxesBuilder(session).build(), session))
+                .concatMap(mailboxMetadata -> Mono.fromCallable(() -> mailboxManager.getMailbox(mailboxMetadata.getId(), session)))
+                .concatMap(Throwing.function(messageManager -> correctMailboxProjectionItems(progress, messageManager, session)))
+                .doOnComplete(progress.processedUserCount::incrementAndGet)
+                .onErrorContinue((error, o) -> {
+                    LOGGER.error("JMAP preview re-computation aborted for {}", username, error);
+                    progress.failedUserCount.incrementAndGet();
+                })
+                .then();
+        } catch (MailboxException e) {
+            LOGGER.error("JMAP preview re-computation aborted for {} as we failed listing user mailboxes", username, e);
+            progress.failedUserCount.incrementAndGet();
+            return Mono.empty();
+        }
+    }
+
+    private Mono<Void> correctMailboxProjectionItems(Progress progress, MessageManager messageManager, MailboxSession session) throws MailboxException {
+        return Iterators.toFlux(messageManager.getMessages(MessageRange.all(), FetchGroup.MINIMAL, session))
+            .concatMap(Throwing.function(messageResult -> Iterators.toFlux(messageManager.getMessages(MessageRange.all(), FetchGroup.BODY_CONTENT, session))))
+            .map(Throwing.function(messageResult -> Pair.of(messageResult.getMessageId(), projectionItemFactory.from(messageResult))))
+            .concatMap(pair -> Mono.from(messageFastViewProjection.store(pair.getKey(), pair.getValue()))
+                .doOnSuccess(any -> progress.processedMessageCount.incrementAndGet()))
+            .onErrorContinue((error, triggeringValue) -> {
+                LOGGER.error("JMAP preview re-computation aborted for {} - {}", session.getUser(), triggeringValue, error);
+                progress.failedMessageCount.incrementAndGet();
+            })
+            .then();
+    }
```

Comment:
there are two many things happening here IMO (and many are just technical details).
Could you extract things to get:
```
return listMessagesX
            .concatMap(messages -> retrieveBodies(messages))
            .map(completeMessages -> asIndexedProjectionItem(completeMessages))
            .concatMap(projection -> store(projection)
                .doOnSuccess(any -> incrementSuccessCounter())
            .onErrorContinue(error -> incrementFailureCounter(error))
            .then();
```